vllm/v1/metrics/perf.py

```diff
@@ -0,0 +1,898 @@
+
+    def add_parser(self, parser: Parser) -> None:
+        self.parsers.append(parser)
+
+    def parse(self, vllm_config: VllmConfig) -> ParsedArgs:
+        args = ParsedArgs()
+        for parser in self.parsers:
+            args = parser.parse(args, vllm_config)
+        return args
+
+
+_COMPONENT_METRICS_REGISTRY: dict[str, type["ComponentMetrics"]] = {}
+
+
+class ComponentMetrics(BaseModel, ABC):
+    """
+    Each concrete ComponentMetrics class is associated with:
+    - fields that are required for metric derivation
+      (fields are specified/validated through pydantic model)
+    - parser to parse VllmConfig into fields
+    - metric methods that derive flops/bytes for a given execution context
+    """
+
+
+    @classmethod
+    @abstractmethod
+    def component_type(cls) -> str: ...
+
+    @classmethod
+    @abstractmethod
+    def get_parser(cls) -> ParserChain: ...
+
+    def __init_subclass__(cls):
+        _COMPONENT_METRICS_REGISTRY[cls.component_type()] = cls
+
+    @classmethod
+    def from_vllm_config(cls, vllm_config: VllmConfig) -> Self:
+        """
+        Instantiate this class from VllmConfig.
+        Raises ValueError if parsing fails.
+        """
+
+        parser = cls.get_parser()
+        parsed_args = parser.parse(vllm_config)
+        return cls.model_validate(parsed_args.model_dump())
+
+    @classmethod
+    def registered_metrics(cls) -> Iterable[type["ComponentMetrics"]]:
+        return iter(_COMPONENT_METRICS_REGISTRY.values())
+
+    @abstractmethod
+    def get_num_flops_breakdown(self, ctx: ExecutionContext, per_gpu: bool = True) -> Breakdown: ...
+
+    @abstractmethod
+    def get_read_bytes_breakdown(self, ctx: ExecutionContext, per_gpu: bool = True) -> Breakdown: ...
+
+    @abstractmethod
+    def get_write_bytes_breakdown(self, ctx: ExecutionContext, per_gpu: bool = True) -> Breakdown: ...
+
+    def get_num_flops(self, ctx: ExecutionContext, per_gpu: bool = True) -> int:
+        return sum(self.get_num_flops_breakdown(ctx, per_gpu).model_dump().values())
+
+    def get_read_bytes(self, ctx: ExecutionContext, per_gpu: bool = True) -> int:
+        return sum(self.get_read_bytes_breakdown(ctx, per_gpu).model_dump().values())
+
+    def get_write_bytes(self, ctx: ExecutionContext, per_gpu: bool = True) -> int:
+        return sum(self.get_write_bytes_breakdown(ctx, per_gpu).model_dump().values())
+
+
+#### parsers ####
+
+
+class BaseConfigParser(Parser):
+    def parse(self, args: ParsedArgs, vllm_config: VllmConfig) -> ParsedArgs:
+        cfg = vllm_config.model_config.hf_config
+        if hasattr(cfg, "text_config") and cfg.text_config is not None:
+            cfg = cfg.text_config
+
+        args.vocab_size = get_required(cfg, "vocab_size")
+        args.hidden_size = get_required(cfg, "hidden_size")
+        args.num_attention_heads = get_required(cfg, "num_attention_heads")
+        args.num_hidden_layers = get_required(cfg, "num_hidden_layers")
+
```

Comment:
In vLLM, we have some more general way to get these model parameters: https://github.com/vllm-project/vllm/blob/5f3cd7f7f20a8e4445d70cbd1f5475175ef391e3/vllm/config/model.py#L1230-L1353

In the future, we are moving towards a more unified config type to make this parsing easier: https://github.com/vllm-project/vllm/issues/24384